### PR TITLE
WIP Issue #177 - Adding an authentication middleware API

### PR DESCRIPTION
**Pull request by remiguittaut**
Draft of an Authentication middleware API inspired from Http4s.

The `AuthenticationAdvanced` example shows the proposed design.

Missing tests and documentation.

### CODE REVIEW

File: zio-http/src/main/scala/zhttp/http/auth/package.scala [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package zhttp.http
+
+package object auth extends AuthedRequestSyntax {
+  import Partials._
+
+  type HttpAuthenticationMiddleware[-R, +A] =
+    Http[R, HttpError, Request, AuthedRequest[A]]
+
+  object HttpAuthenticationMiddleware {
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/auth/package.scala:9]: Rename to `Auth` it's concise.

File: zio-http/src/main/scala/zhttp/http/auth/package.scala [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+    def ofM[A]: HttpAuthenticationMiddlewareMPartiallyApplied[A] =
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/auth/package.scala:11]:
```suggestion
    def customM[A]: HttpAuthenticationMiddlewareMPartiallyApplied[A] =
```

File: zio-http/src/main/scala/zhttp/http/auth/package.scala [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+      new HttpAuthenticationMiddlewareMPartiallyApplied[A]
+
+    def of[A]: HttpAuthenticationMiddlewarePartiallyApplied[A] =
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/auth/package.scala:14]:
```suggestion
    def custom[A]: HttpAuthenticationMiddlewarePartiallyApplied[A] =
```

File: zio-http/src/main/scala/zhttp/http/auth/Partials.scala [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        else
+          Http.empty
+      }
+  }
+
+  private[auth] final case class HttpAuthenticationMiddlewarePartiallyApplied[A](private val dummy: Boolean = false)
+    extends AnyVal {
+
+    def apply[R, E <: HttpError, A1 >: A](
+      f: Request => Either[E, AuthedRequest[A1]]
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/auth/Partials.scala:39]:
```suggestion
      f: Request => Http[R, E, AuthedRequest[A1]]
```

File: zio-http/src/main/scala/zhttp/http/auth/Partials.scala [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    ): HttpAuthenticationMiddleware[R, A1] = {
+      Http.identity[Request] >>= { request =>
+        f(request).fold(Http.fail, Http.succeed)
+      }
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/auth/Partials.scala:43]:
```suggestion
      Http.collect[Request](pf).flatten
```

File: example/src/main/scala/AuthenticationAdvanced.scala [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        case (usr, pass) if usr.equalsIgnoreCase(user)
+          && pass.equalsIgnoreCase(password) => usr
+      } toRight(HttpError.Forbidden())
+    } yield AuthInfo(authenticated)
+
+  val authMiddleware: AuthMiddleware =
+    HttpAuthenticationMiddleware.ofM { request =>
+      ZIO.fromEither(authenticate(request))
+         .map(AuthedRequest(_, request))
+    }
```
Comment (tusharmath) [inline example/src/main/scala/AuthenticationAdvanced.scala:45]:
```suggestion
    Auth.custom { request =>
       authenticate(request)
         .fold(Http.fail, Http.succeed)
         .flatten
         .map(AuthedRequest(_, request))
    }
```

File: example/src/main/scala/AuthenticationAdvanced.scala [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    HttpAuthenticationMiddleware.of { request =>
+      authenticate(request).map(AuthedRequest(_, request))
+    }
+*/
+
+  val anonymousApp: UHttpApp = Http.collect[Request] {
+    case Method.GET -> Root / "health" => Response.ok
+  }
+
+  val protectedApp = Http.collectAuthed[AuthInfo] {
```
Comment (tusharmath) [inline example/src/main/scala/AuthenticationAdvanced.scala:60]:
```suggestion
  val protectedApp = Http.collect[AuthRequest[AuthInfo]] {
```

File: zio-http/src/main/scala/zhttp/http/auth/package.scala (hunk 3 above)
Comment (tusharmath) [reply zio-http/src/main/scala/zhttp/http/auth/package.scala:14]: Later we can add vanilla support for various kinds of authentication mechanisms.

File: example/src/main/scala/AuthenticationAdvanced.scala [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+      userNamePassword <- decoded.split(":").map(_.trim).toList match {
+        case usr :: pass :: Nil => Right((usr -> pass))
+        case _ => Left(HttpError.Unauthorized())
+      }
+      (user, password) = userNamePassword
+      authenticated <- users.collectFirst {
+        case (usr, pass) if usr.equalsIgnoreCase(user)
+          && pass.equalsIgnoreCase(password) => usr
+      } toRight(HttpError.Forbidden())
+    } yield AuthInfo(authenticated)
```
Comment (tusharmath) [inline example/src/main/scala/AuthenticationAdvanced.scala:39]: Do we need such an example for auth in user-land code? It's a little intimidating for first-timers 🙈

Comment (remiguittaut) [reply example/src/main/scala/AuthenticationAdvanced.scala:39]: yeah we can simplify, but it's just a simple basic auth :). I'll simplify.

**Comment by remiguittaut** [comment]
Took note of all the remarks, I'll take care of it later. Thanks for taking the time to review :).

File: zio-http/src/main/scala/zhttp/http/auth/Partials.scala (hunk 4 above)
Comment (remiguittaut) [reply zio-http/src/main/scala/zhttp/http/auth/Partials.scala:39]: It's missing the `A` type parameter in `Http` if I do that, do  you mean replacing `Request => Either[E, AuthedRequest[A1]]` by `Http[E, Request, AuthedRequest[A1]]`?

**Comment by remiguittaut** [comment]
Hi, added a new commit according to some of your comments. (Concise naming and example simplification). I also simplified the Api surface so the user doesn't have to use the type `AuthedRequest` at all in his auth middleware implementation. I stand however behind keeping `collectAuthed` and shielding the user from using the almost internal type `Http` in his middleware. It makes it very easy for a new user to plug-in his code in almost vanilla scala (the only zio-http specific type to handle is `HttpError` in the error channel of the middleware implementation).

File: zio-http/src/main/scala/zhttp/http/auth/Partials.scala (hunk 4 above)
Comment (tusharmath) [reply zio-http/src/main/scala/zhttp/http/auth/Partials.scala]: Fixed the comment.

**Comment by tusharmath** [comment]
This is fixed in RC 18

**Pull request closed without merging**